File: src/day1p2.ts
import { getFileText } from "./utils/get-file-text";

const fileText = await getFileText(import.meta.dir + "/day1.input");

console.time("result gen");

const replacementMap = new Map([
  ["one", "one1one"],
  ["two", "two2two"],
  ["three", "three3three"],
  ["four", "four4four"],
  ["five", "five5five"],
  ["six", "six6six"],
  ["seven", "seven7seven"],
  ["eight", "eight8eight"],
  ["nine", "nine9nine"],
]);

const result = fileText
  .trim()
  .split("\n")
  .map((value) => {
    replacementMap.forEach(
      (newVal, oldVal) => (value = value.replaceAll(oldVal, newVal)),
    );
    return value;
  })
  .reduce((acc, value) => {
    let firstNumChar: string | undefined = undefined;
    let lastNumChar: string | undefined = undefined;

    for (let index = 0; index < value.length; index++) {
      if (!firstNumChar && !isNaN(parseInt(value[index]))) {
        firstNumChar = value[index];
      }

      if (!lastNumChar && !isNaN(parseInt(value[value.length - 1 - index]))) {
        lastNumChar = value[value.length - 1 - index];
      }

      if (firstNumChar && lastNumChar) {
        break;
      }
    }

    return String(Number(acc) + Number(`${firstNumChar}${lastNumChar}`));
  }, "0");

console.timeEnd("result gen");
console.log(result);

File: src/day1p2.rs
use std::io;

use crate::get_file_text;

pub fn solve_day_1_part_2() -> io::Result<()> {
    let text = get_file_text("src/day1.input")?;

    let original_text = String::from(text);

    let replaced_text = original_text
        .replace("one", "o1e")
        .replace("two", "t2o")
        .replace("three", "t3e")
        .replace("four", "f4r")
        .replace("five", "f5e")
        .replace("six", "s6x")
        .replace("seven", "s7n")
        .replace("eight", "e8t")
        .replace("nine", "n9e");

    let rows = replaced_text.trim().split("\n");

    let mut sum = 0;

    for row in rows {
        let mut first_num_char: char = '_';
        let mut last_num_char: char = '_';

        for char in row.chars() {
            let convert_result = char.to_string().parse::<i32>();

            match convert_result {
                Ok(_) => {
                    first_num_char = char;
                    break;
                }
                Err(_) => continue,
            }
        }

        for char in row.chars().rev() {
            let convert_result = char.to_string().parse::<i32>();

            match convert_result {
                Ok(_) => {
                    last_num_char = char;
                    break;
                }
                Err(_) => continue,
            }
        }

        match format!("{}{}", first_num_char, last_num_char).parse::<i32>() {
            Ok(val) => sum += val,
            Err(_) => println!("Could not parse number!"),
        };
    }

    println!("Day_1_Part_2: {}", sum);

    return Ok(());
}
